File: android/app/src/main/java/com/nextpage/data/remote/addons/CuratedCatalogProvider.kt
package com.nextpage.data.remote.addons

import android.content.Context
import com.nextpage.data.remote.catalog.CatalogBook
import com.nextpage.data.remote.catalog.CatalogErrorCode
import com.nextpage.data.remote.catalog.CatalogProvider
import com.nextpage.data.remote.catalog.CatalogSourceInfo
import com.nextpage.data.remote.catalog.CatalogSourceKind
import com.nextpage.data.remote.catalog.PagedResult
import com.nextpage.data.remote.catalog.parseCatalogSource
import com.nextpage.data.remote.catalog.resolveDownloadUrl
import org.json.JSONArray
import org.json.JSONObject

/**
 * Curated first-party bundle: read-only, manifest-shaped `assets/addons/curated.json`
 * (byte-identical to desktop `services/addons/curated.json`). Every entry passes
 * [ManifestValidator] at construction — a malformed bundle fails fast. Browse-only:
 * search contributes nothing to the composite (empty page, no I/O, keeping the
 * zero-addon parity byte-for-byte) and getDetails fails stable NOT_FOUND
 * ("not routable"). Never a registry row. Mirrors desktop `CuratedCatalogProvider.ts`.
 */
class CuratedCatalogProvider private constructor(
    private val sources: List<CatalogSourceInfo>
) : CatalogProvider {

    constructor(context: Context) : this(loadSources(context))

    /** Test/fixture entry: parse + validate a bundle payload without assets. */
    constructor(bundleJson: String) : this(CuratedCatalogProvider.parseBundle(bundleJson))

    override fun listSources(): List<CatalogSourceInfo> = sources

    /** Browse-only: contributes nothing to composite search, never I/O. */
    override suspend fun search(query: String, page: Int): PagedResult =
        PagedResult(emptyList(), null, 0)

    /** Curated sources are not routable; stable NOT_FOUND, no I/O. */
    override suspend fun getDetails(id: String): CatalogBook {
        throw com.nextpage.data.remote.catalog.catalogError(
            CatalogErrorCode.NOT_FOUND,
            "curated sources are not routable: $id"
        )
    }

    override fun resolveDownloadUrl(formats: Map<String, String>, preferEpub: Boolean): String =
        resolveDownloadUrl(formats, preferEpub)

    companion object {
        private const val ASSET_PATH = "addons/curated.json"

        private fun loadSources(context: Context): List<CatalogSourceInfo> {
            val json = context.assets.open(ASSET_PATH).bufferedReader().use { it.readText() }
            return parseBundle(json)
        }

        /** Visible for tests: parse + validate a bundle payload without assets. */
        fun parseBundle(json: String): List<CatalogSourceInfo> {
            val root = JSONObject(json)
            val addons = root.optJSONArray("addons")
                ?: throw AddonFetchException(AddonFetchErrorCode.INVALID_MANIFEST, "curated bundle must be {addons: [...]}")
            val out = mutableListOf<CatalogSourceInfo>()
            for (i in 0 until addons.length()) {
                val entry = addons.opt(i) as? JSONObject
                    ?: throw AddonFetchException(AddonFetchErrorCode.INVALID_MANIFEST, "curated entry must be an object")
                // Shared validation gate: same rules as registry installs.
                ManifestValidator.validate(entry.toString().toByteArray(Charsets.UTF_8), "application/json")
                val manifestId = entry.getString("id")
                out.add(
                    CatalogSourceInfo(
                        sourceId = parseCatalogSource("builtin:$manifestId"),
                        name = entry.getString("name"),
                        kind = CatalogSourceKind.CURATED
                    )
                )
            }
            return out
        }
    }
}

File: desktop/src/lib/shared/services/addons/CuratedCatalogProvider.ts
/**
 * CuratedCatalogProvider — read-only, manifest-shaped first-party bundle.
 * The bundled curated.json lists exactly six public-domain sources; every
 * entry passes the shared manifest validation at construction (a bad bundle
 * fails fast, guarded by tests). Browse-only: it never serves search results
 * (search returns an empty page without I/O, keeping the composite's
 * zero-addon parity byte-for-byte) and getDetails fails stable NOT_FOUND
 * ("not routable"). Never a registry row.
 */
import { catalogError } from '../catalog/errors';
import type {
  CatalogBook,
  CatalogProvider,
  CatalogSourceInfo,
  PagedResult,
} from '../catalog/CatalogProvider';
import { parseCatalogSource } from '../catalog/CatalogProvider';
import { resolveDownloadUrl } from '../catalog/mappers';
import { validateManifest, type AddonManifest } from '@nextpage/manifest-validator';
import curatedJson from './curated.json';

interface CuratedBundle {
  addons: unknown[];
}

function loadCuratedManifests(): AddonManifest[] {
  const bundle = curatedJson as CuratedBundle;
  if (!Array.isArray(bundle.addons)) {
    throw new Error('curated bundle must be {addons: [...]}');
  }
  const encoder = new TextEncoder();
  return bundle.addons.map((entry) =>
    validateManifest(encoder.encode(JSON.stringify(entry)), 'application/json'),
  );
}

const EMPTY_PAGE: PagedResult = { results: [], nextPage: null, totalCount: 0 };

export class CuratedCatalogProvider implements CatalogProvider {
  private readonly sources: CatalogSourceInfo[];

  constructor() {
    this.sources = loadCuratedManifests().map((manifest) => ({
      sourceId: parseCatalogSource(`builtin:${manifest.id}`),
      name: manifest.name,
      kind: 'curated' as const,
    }));
  }

  listSources(): CatalogSourceInfo[] {
    return this.sources;
  }

  /** Browse-only: contributes nothing to composite search, never I/O. */
  async search(_query: string, _page: number): Promise<PagedResult> {
    return EMPTY_PAGE;
  }

  /** Curated sources are not routable; stable NOT_FOUND, no I/O. */
  async getDetails(id: string): Promise<CatalogBook> {
    throw catalogError('NOT_FOUND', `curated sources are not routable: ${id}`);
  }

  resolveDownloadUrl(formats: Record<string, string>, preferEpub: boolean): string {
    return resolveDownloadUrl(formats, preferEpub);
  }
}
